Add support for additional disks in VSphereMachineTemplate

dgmorales opened the pull request:
### PR Description
Add support for additional disks in VSphereMachineTemplate, so we can  add extra disks to nodes created by Cluster API on VSphere.

It does that by supporting the additionalDisksGiB attribute, but omitting it in the rendered object if value is unset.

This is because VsphereMachineTemplates are immutable objects. So adding this attribute set to an empty list on an existent object triggers an error on Kubernetes, because it is considered a change in the spec.

This will trigger errors for existing VsphereMachineTemplates that don't need extra disks, and have no change at all, causing us trouble. So instead of adding that empty default, we omit the attribute when unset, and document its existence.

### Checklist

 - [x] Have you reviewed and updated the chart default values if necessary?
 - [x] Have you reviewed and updated the chart documentation if necessary?
 - [x] Does your branch follow the naming convention of `{chartNameWithDashes}-v{versionString}-{optionalPatchVersion}`?
 - [x] Have you bumped the version in the chart's `Chart.yaml`?


### Tests

I have tested this version together with the changes on https://github.com/powerhome/pacman/pull/190. It is being successfully used by this ArgoCD application:
https://cd.mc.prod.hq.powerapp.cloud/applications/argo/clusterapi-wc-alpha-gm

dgmorales committed:
feat: Add support for setting additional disks on a VSphereMachineTemplate

dgmorales commented:
I will keep this in draft state until I perform a test on or alpha cluster testing that syntax on a real vsphere machine template. But it should be good.

dgmorales committed:
feat: Add support for adding files only on kubeadmConfigTemplates

Current rendering logic inserts inline files on both kubeadmControlPlane objects (used for control plane nodes) and kubeadmConfigTemplates objects (used by worker nodes). We want to be able to add files just to worker nodes.

iblackman commented on charts/clusterapi-resources/values.yaml:
it could be interesting to add an example to the values file of the new parameter `kubeadmConfigSpec.kubeadmConfigTemplateFiles` similar to https://github.com/powerhome/helm-charts/blob/464918e8da8008366995dc722d79a2769c85e456/charts/clusterapi-resources/values.yaml#L101 since they have different purposes, maybe we can add some comment regarding them

dgmorales replied:
Good point. See my new updates on that line.

dgmorales committed:
Add example usage of kubeadmConfigSpec.kubeadmConfigTemplateFiles

dgmorales committed:
Make additionalDisksGiB default to be unset.

VsphereMachineTemplates are immutable objects. So even adding this attribute set to an empty list triggers an error on Kubernetes, because it is considered a change in the spec.

This will trigger errors for existing VsphereMachineTemplates that don't need extra disks, and have no change at all.

So instead of adding that empty default, we omit the attribute when unset, and document its existence.

dgmorales commented on charts/clusterapi-resources/templates/KubeadmConfigTemplate.yaml:
```diff
@@ -26,6 +26,9 @@ spec:
       files:
       {{- toYaml . | nindent 6 }}
 {{- end }}
+{{- with $.Values.kubeadmConfigSpec.kubeadmConfigTemplateFiles }}
+      {{- toYaml . | nindent 6 }}
+{{- end }}
```
I've changed the approach here, to avoid this silly diff in the image:

![Screenshot 2025-04-25 at 15 01 18](https://github.com/user-attachments/assets/82bd162e-3e5e-42a6-9c88-1492e75f2a71)

This triggers errors on ArgoCD because of immutable objects for the control plane (which needs no change In disks, so it's just causing trouble). 

Without this, we will needlessly have to rollout new control nodes as well.

iblackman replied:
I think you meant to comment on  https://github.com/powerhome/helm-charts/pull/53/files#diff-e0d6daef3a5844d7380d61c36776b62dbf8d30b452becb9122c899b71eddb0e1R17-R19 but I understood haha
